code/datums/components/trader/trader.dm

```diff
@@ -0,0 +1,413 @@
+#define TRADER_RADIAL_OUT_OF_STOCK "TRADER_RADIAL_OUT_OF_STOCK"
+#define TRADER_RADIAL_DISCUSS_BUY "TRADER_RADIAL_DISCUSS_BUY"
+#define TRADER_RADIAL_DISCUSS_SELL "TRADER_RADIAL_DISCUSS_SELL"
+
+//The defines below show the index the info is located in the product_info entry list
+
+#define TRADER_PRODUCT_INFO_PRICE 1
+#define TRADER_PRODUCT_INFO_QUANTITY 2
+//Only valid for wanted_items
+#define TRADER_PRODUCT_INFO_PRICE_MOD_DESCRIPTION 3
+
+/**
+ * # Trader NPC Component
+ * Manages the barks and the stocks of the traders
+ * Also manages the interactive radial menu
+ */
+/datum/component/trader
+
+	/**
+	 * Format; list(TYPEPATH = list(PRICE, QUANTITY))
+	 * Associated list of items the NPC sells with how much they cost and the quantity available before a restock
+	 * This list is filled by Initialize(), if you want to change the starting products, modify initial_products()
+	 * *
+	 */
+	var/list/products = list()
+	/**
+	 * A list of wanted items that the trader would wish to buy, each typepath has a assigned value, quantity and additional flavor text
+	 *
+	 * CHILDREN OF TYPEPATHS INCLUDED IN WANTED_ITEMS WILL BE TREATED AS THE PARENT IF NO ENTRY EXISTS FOR THE CHILDREN
+	 *
+	 * As an additional note; if you include multiple children of a typepath; the typepath with the most children should be placed after all other typepaths
+	 * Bad; list(/obj/item/milk = list(100, 1, ""), /obj/item/milk/small = list(50, 2, ""))
+	 * Good; list(/obj/item/milk/small = list(50, 2, ""), /obj/item/milk = list(100, 1, ""))
+	 * This is mainly because sell_item() uses a istype(item_being_sold, item_in_entry) to determine what parent should the child be automatically considered as
+	 * If /obj/item/milk/small/spooky was being sold; /obj/item/milk/small would be the first to check against rather than /obj/item/milk
+	 *
+	 * Format; list(TYPEPATH = list(PRICE, QUANTITY, ADDITIONAL_DESCRIPTION))
+	 * Associated list of items able to be sold to the NPC with the money given for them.
+	 * The price given should be the "base" price; any price manipulation based on variables should be done with apply_sell_price_mods()
+	 * ADDITIONAL_DESCRIPTION is any additional text added to explain how the variables of the item effect the price; if it's stack based, it's final price depends how much is in the stack
+	 * EX; /obj/item/stack/sheet/mineral/diamond = list(500, INFINITY, ", per 100 cm3 sheet of diamond")
+	 * This list is filled by Initialize(), if you want to change the starting wanted items, modify initial_wanteds()
+	*/
+	var/list/wanted_items = list()
+
+	///Contains images of all radial icons
+	var/static/list/radial_icons_cache = list()
+
+	///Contains information of a specific trader
+	var/datum/trader_data/trader_data
+
+/datum/component/trader/Initialize(trader_data_path)
+	. = ..()
+	if (!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/mob/living/trader = parent
+	if (!trader.ai_controller)
+		return COMPONENT_INCOMPATIBLE
```

Comment:
```suggestion
	if (isnull(trader.ai_controller))
		return COMPONENT_INCOMPATIBLE
```

Reply:
Is this also actually true (we're incompatible if we have no ai controller)?
It seems like the component's behaviour _should_ work just fine in the absence of one, if you wanted a completely stationary vendor.

It does not even look like the _living_ requirement comes up very much and you might be able to make that optional as well, such that this would be more of a two-way vending machine component in some ways.

Reply:
I'll remove the ai controller bit for now, will return to se if everything works fine if i remove the living requirement.